README
HTML-Auto

This module attempts to provide easy access to some typical structures
used in HTML, for example matrices. The module is still in heavy
development.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc HTML::Auto

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTML-Auto

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/HTML-Auto

    CPAN Ratings
        http://cpanratings.perl.org/d/HTML-Auto

    Search CPAN
        http://search.cpan.org/dist/HTML-Auto/



Changes
Revision history for HTML-Auto

{{$NEXT}}

0.09      2016-07-18 20:51:13+01:00 Europe/Lisbon
          * Small documentation updates.
          * Fix white spaces.
          * dzil updates.

0.08      2016-07-18 16:42:52+01:00 Europe/Lisbon
          * Move to dzil.

0.07      Thu Jul 14 15:46:33 WEST 2016
          * Update default CSS.
          * Add options to the matrix function: diagonal, ucfirst, and format.

0.06      Thu Nov 12 20:55:02 WET 2015
          Handle paths correctly in Windows too (reported by CHORNY).

0.05      Sun Nov  8 17:27:23 WET 2015
          Fix tests to run in recent perl releases (reported by SREZIC).

0.04      Tue Jun  5 16:02:52 WEST 2012
          Fixed UTF-8 in POD

0.03      Tue May 22 22:50:31 WEST 2012
          Cells can have a hover pop-up.
		  Additional CSS can be provided.
		  Added some tests.

0.02      Sat Apr 21 14:53:52 WEST 2012
          Update docs.

0.01      Sun Sep 25 16:45:04 WEST 2011
          First version.

0.01_3    Tue Sep 20 23:24:32 WEST 2011
          Matrix elements may be scalars.

0.01_2    Sun Jan 30 13:58:32 WET 2011
          Fix makefile dependencies.

0.01_1    Fri Jan 28 14:11:09 WET 2011
          First devel release.


lib/HTML/Auto.pm
package HTML::Auto;
# ABSTRACT: write HTML for common elements

use warnings;
use strict;

use Template;
use HTML::Auto::Templates;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(matrix h v);

sub matrix {
  my ($cols,$lines,$data,$options) = @_;

  if ($options->{ucfirst}) {
    foreach (@$cols, @$lines) {
      $_ = ucfirst($_);
    }
  }

  my $vals = [];
  my $attrs = [];
  my $more = [];

  foreach my $row (@$data) {
    my $vrow = [];
    my $arow = [];
    my $mrow = [];
    foreach(@$row){
      if (ref($_)){
        push @$vrow, $_->{v};
        push @$arow, $_->{a};
        push @$mrow, $_->{more_info};
      }
      else {
        push @$vrow, $_;
        push @$arow, undef;
        push @$mrow, undef;
      }
    }
    push @$vals, $vrow;
    push @$attrs, $arow;
    push @$more, $mrow;
  }

  my $vars = {
      cols  => $cols,
      lines => $lines,
      vals  => $vals,
      attrs => $attrs,
      more  => $more,
    };
  $vars->{css} = $options->{css}
    if $options->{css};
  $vars->{myformat} = $options->{format}
    if $options->{format};
  $vars->{diagonal} = $options->{diagonal}
    if $options->{diagonal};

  my $template_name = 'matrix';
  __process($template_name, $vars);
}

sub h {
  my (@list) = @_;

  my $vars = {
      list => [@list],
    };
  my $template_name = 'h';

  __process($template_name, $vars);
}

sub v {
  my (@list) = @_;

  my $vars = {
      list => [@list],
  };
  my $template_name = 'v';

  __process($template_name, $vars);
}

sub __process {
  my ($template_name,$vars) = @_;

  # build html from template
  my $template_config = {
      INCLUDE_PATH => [ 'templates' ],
    };
  my $template = Template->new({
      LOAD_TEMPLATES => [ HTML::Auto::Templates->new($template_config) ],
    });
  my $html;
  $template->process($template_name, $vars, \$html);

  return $html;
}

1;

__END__

=encoding utf-8

=head1 SYNOPSIS

Simple example:

  use HTML::Auto qw/matrix h v/;

  my @cols = qw/c1 c2 c3 c4 c5/;
  my @lines = qw/l1 l2 l3 l4 l5/;
  my $data =
     [ [1,2,3,4,5],
       [6,7,8,9,0],
       [1,1,1,1,1],
       [2,2,2,2,2],
       [3,3,3,3,3] ];

  my $m = matrix(\@cols,\@lines,$data);

  print v(
          h($m,$m,$m),
          h($m,$m),
        );

Using attributes:

  use HTML::Auto qw/matrix h v/;

  my @cols = qw/c1 c2/;
  my @lines = qw/l1 l2/;
  my $data =
     [
       [
         {v => 1, a => { style => 'background: green'}},
         2
       ],
       [
         {v => 3, a => {class => 'foo'}},
         {v => 4, a => {style => 'color: red'}}
       ]
     ];

  my $m = matrix(\@cols,\@lines,$data);

  print v(
          h($m)
        );

With mouse-over span:

  use HTML::Auto qw/matrix h v/;

  my @cols = qw/c1 c2/;
  my @lines = qw/l1 l2/;
  my $data =
     [[1,2],
    [3,
    { v=> 4,
      more_info => "This is a pop-up!"
    }]
   ];


  my $m = matrix(\@cols,\@lines,$data);

  print v(
          h($m)
        );

Passing additional CSS:

  use HTML::Auto qw/matrix h v/;

  my @cols = qw/c1 c2/;
  my @lines = qw/l1 l2/;
  my $data =
     [
       [
         {v => 1, a => { class => 'warn'}},
         2
       ],
       [3,4]
     ];

  my $options = { css => '.warn { background-color: yellow !important; }' };

  my $m = matrix(\@cols,\@lines,$data,$options);

  print v(
          h($m)
        );

=func matrix

Build a matrix. Some options are available to pass to the matrix function:

=over 6

=item C<diagonal>

Highlight the diagonal of the matrix.

  my $m = matrix(\@cols,\@lines,$data, {diagonal => 1});

=item C<format>

Pass a string to be used by the C<format> filter in the TT2 template.

  my $m = matrix(\@cols,\@lines,$data, {format => '%.6f'});

=item C<ucfirst>

Option to uppercase first letter in columns and lines labels.

  my $m = matrix(\@cols,\@lines,$data, {ucfirst => 1});

=back

=func h

A function to allow horizontal composition.

=func v

A function to allow vertical composition.


lib/HTML/Auto/Templates.pm
package HTML::Auto::Templates;
# ABSTRACT: stores HTML::Auto templates

use base qw/Template::Provider/;

use warnings;
use strict;

use File::Basename;

my $templates = {

'matrix' => <<'EOT'
<style type="text/css">
span.vertical {
    -webkit-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    transform: rotate(180deg);
    writing-mode: tb-rl;
    filter: flipv fliph;
    display: block;
    width: 20px;
    white-space: nowrap;
}

table.auto th {
    padding-top: 24px;
    padding-bottom: 10px;
    padding-left: 5px;
    padding-right: 5px;
    width: 20px;
    background-color: #aaaaaa;
}

table.auto td {
    text-align: center;
    width: 30px;
    background-color: #eeeeee;
    padding: 6px;
}

table.auto td.fst {
    width: 80px;
    font-weight: bold;
    background-color: #aaaaaa;
    padding: 5px;
}

th.empty {
    background-color: white !important;
}

table.auto td.mid {
    background-color: #cccccc;
}

td:hover {
    background-color: #aaaaaa;
}

td.more_info { 
    position:relative;
    z-index:24;
    text-decoration:none;
    cursor: default;
    color: black;
    width: 80px;
}

td.more_info:hover{
    z-index:25;
}

td.more_info td {
    width: auto;
}

td.more_info span {
    display: none;
}

td.more_info:hover span { 
    display:block;
    position:absolute;
    border:1px solid #ccc;
    min-width:24em;
    background-color:#fff;
    color:#000;
    text-align: left;
    font-size: 80%;
    text-decoration: none;
} 

[% IF css %]
[% css %]
[% END %]
</style>

<table class="auto">
  <tr>
    <th class="empty"> </th>
      [% FOREACH i IN cols %]
        <th> <span class="vertical">[% i -%]</span></th>
      [% END %]
  </tr>
  [% i_c = 0 %]
  [% FOREACH i IN vals %]
    <tr>
      <td class="fst">[% lines.shift -%]</td>
        [% j_c = 0 %]
        [% FOREACH j IN i %]
          <td
            [% class = "" %]
            [% IF diagonal AND i_c == j_c %]
              [% class = "mid" %]
            [% END %]
            [% IF more.$i_c.$j_c %]
              [% class = "more_info " _  class %]
            [% END %]
            [% IF attrs.$i_c.$j_c.class %]
              [% class = class _ " " _ attrs.$i_c.$j_c.class %]
            [% END %]
            [% attrs.$i_c.$j_c.delete('class') %]
            [% IF class.length != 0 %] class="[% class %]" [% END %]
            [% FOREACH att IN attrs.$i_c.$j_c.keys %]
              [% att %]="[% attrs.$i_c.$j_c.$att %]" 
            [% END %]
            >[% IF myformat %][% j | format(myformat) %][% ELSE %][% j %][% END %]
              [% IF more.$i_c.$j_c %]
                <span>[% more.$i_c.$j_c %]</span> 
              [% END %]
          </td> 
	      [% j_c = j_c + 1 %]
        [% END %]
      </tr>
      [% i_c = i_c + 1 %]
  [% END %]
</table>
EOT
,
'h' => <<'EOT'
<div>
  [% FOREACH i IN list %]
    <div style="float: left;">[% i %]</div>
  [% END %]
  <span style="clear: both;"></div>
</div>
EOT
,
'v' => <<'EOT'
[% FOREACH i IN list %]
  <div>[% i %]</div>
[% END %]
EOT
};

sub _template_modified {
  my($self,$path) = @_;

  return 1;
}

sub _template_content {
  my ($self, $path) = @_;

  $path = basename($path);;
  $self->debug("get $path") if $self->{DEBUG};

  my $data = $templates->{$path};
  my $error = "error: $path not found";
  my $mod_date = 1;

  return $data;
}

1;

